lib/logstash/outputs/logstash.rb

```diff
@@ -61,129 +59,171 @@ def initialize(*a)
       fail LogStash::ConfigurationError, 'The `logstash` output does not have an externally-configurable `codec`'
     end
 
-    if @ssl_certificate_authorities && @ssl_certificate_authorities.size > 1
-      fail LogStash::ConfigurationError, 'The `logstash` output supports at most one `ssl_certificate_authorities` path'
+    @headers = {}.tap do | header |
+      header["Content-Type"] = "application/x-ndjson"
+      header["Content-Encoding"] = "gzip"
     end
 
-    logger.debug("initializing inner HTTP output plugin")
-    @internal_http = plugin_factory.output('http').new(inner_http_output_options)
-    logger.debug("inner HTTP output plugin has been initialized")
+    logger.debug("`logstash` output plugin has been initialized.")
   end
 
   def register
-    logger.debug("registering inner HTTP output plugin")
-    @internal_http.register
-    logger.debug("inner HTTP output plugin has been registered")
+    logger.debug("Registering `logstash` output plugin.")
+
+    @username = normalize_config(:username) do |normalize|
+      normalize.with_deprecated_alias(:user)
+    end
+
+    if @ssl_enabled == false
+      rejected_ssl_settings = @original_params.keys.select { |k| k.start_with?('ssl_') } - %w(ssl_enabled)
+      fail(LogStash::ConfigurationError, "Explicit SSL-related settings not supported because `ssl_enabled => false`: #{rejected_ssl_settings}") if rejected_ssl_settings.any?
+    end
+
+    # if we don't initialize now, we get runtime error when sending events if there are issues with configs
+    @http_client = client
+    @load_balancer = LoadBalancer.new(construct_host_uri)
+
+    logger.debug("`logstash` output plugin has been registered.")
   end
 
   def multi_receive(events)
     return if events.empty?
-    logger.trace("proxying #{events.size} events to inner HTTP plugin")
-    @internal_http.multi_receive(events)
-  rescue => e
-    logger.error("inner HTTP plugin has had an unrecoverable exception: #{e.message} at #{e.backtrace.first}")
-    raise
+
+    send_events(events)
   end
 
   def stop
-    logger.debug("stopping inner HTTP output plugin")
-    @internal_http.stop
-    logger.debug('inner HTTP output plugin has been stopped')
+    logger.debug("`logstash` output plugin has been stopped.")
   end
 
   def close
-    logger.debug("closing inner HTTP output plugin")
-    @internal_http.close
-    logger.debug('inner HTTP output plugin has been closed')
+    logger.debug("Closing `logstash` output plugin.")
+    http_client.close
+    logger.debug("`logstash` output plugin has been closed.")
   end
 
-  DEFAULT_PORT = 9800.freeze
+  def pipeline_shutdown_requested?
+    return super if defined?(super) # since LS 8.1.0
+    execution_context&.pipeline&.shutdown_requested?
+  end
+
+  def abort_batch_if_available!
+    raise org.logstash.execution.AbortedBatchException.new if abort_batch_present?
+  end
```

Comment:
Indeed, runtime `::Gem::Version.create(LOGSTASH_VERSION) >= ::Gem::Version.create('8.8.0')` comparison is expensive, let me finalize from your suggestions.

Reply:
I preferred to stick to versions, since those changes (shutdown request and batch abort features) are obvious. Applied with [recent commit](https://github.com/logstash-plugins/logstash-integration-logstash/pull/16/commits/dcd76b1f38fdf61ab89ef7f7fe49a313c3fba39a) after multiple tests on 7.17.14, 8.0, 8.10.4 and 8.11 (local, upstream main) and let's take an offline to discuss if anything we can improve.
Thank you!